contracts/test/LONStaking.t.sol

```diff
@@ -433,80 +471,302 @@ contract LONStakingTest is Test {
 
-    function testRedeemPartial() public {
-        _stake(user, DEFAULT_STAKE_AMOUNT);
-        vm.startPrank(user);
-        lonStaking.unstake();
-
-        vm.warp(block.timestamp + COOLDOWN_SECONDS + 1);
-
+    function _redeemAndValidate(address redeemer, uint256 redeemAmount) internal {
+        if (redeemAmount > lonStaking.balanceOf(redeemer)) redeemAmount = lonStaking.balanceOf(redeemer);
+        bool redeemPartial = lonStaking.balanceOf(redeemer) > redeemAmount;
         uint256 lonStakingXLONBefore = lonStaking.totalSupply();
         BalanceSnapshot.Snapshot memory lonStakingLON = BalanceSnapshot.take(address(lonStaking), address(lon));
-        BalanceSnapshot.Snapshot memory userXLON = BalanceSnapshot.take(user, address(lonStaking));
-        BalanceSnapshot.Snapshot memory userLON = BalanceSnapshot.take(user, address(lon));
+        BalanceSnapshot.Snapshot memory redeemerXLON = BalanceSnapshot.take(redeemer, address(lonStaking));
+        BalanceSnapshot.Snapshot memory redeemerLON = BalanceSnapshot.take(redeemer, address(lon));
 
-        uint256 redeemAmount = lonStaking.balanceOf(user).div(2);
         uint256 expectedLONAmount = _getExpectedLONWithoutPenalty(redeemAmount);
+        vm.prank(redeemer);
         lonStaking.redeem(redeemAmount);
-        vm.stopPrank();
 
         uint256 lonStakingXLONAfter = lonStaking.totalSupply();
         assertEq(lonStakingXLONAfter, lonStakingXLONBefore - redeemAmount);
         lonStakingLON.assertChange(-int256(expectedLONAmount));
-        userXLON.assertChange(-int256(redeemAmount));
-        userLON.assertChange(int256(expectedLONAmount));
-        assertGt(lonStaking.stakersCooldowns(user), 0, "Cooldown record remains until user redeem all shares");
+        redeemerXLON.assertChange(-int256(redeemAmount));
+        redeemerLON.assertChange(int256(expectedLONAmount));
+        if (redeemPartial) {
+            assertGt(lonStaking.stakersCooldowns(redeemer), 0, "Cooldown record remains until user redeem all shares");
+        }
+    }
+
+    function testRedeemPartial() public {
+        _stake(user, DEFAULT_STAKE_AMOUNT);
+        vm.prank(user);
+        lonStaking.unstake();
+
+        vm.warp(block.timestamp + COOLDOWN_SECONDS + 1);
+
+        uint256 redeemAmount = lonStaking.balanceOf(user).div(2);
+        _redeemAndValidate(user, redeemAmount);
+    }
+
+    function testFuzz_RedeemPartial(uint256 stakeAmount, uint256 redeemAmount) public {
+        vm.assume(stakeAmount > 0);
+        vm.assume(redeemAmount > 0);
+        vm.assume(redeemAmount < stakeAmount);
+        vm.assume(stakeAmount <= lon.cap());
+        vm.assume(stakeAmount.add(lon.totalSupply()) <= lon.cap());
+
+        _stake(user, stakeAmount);
+        vm.prank(user);
+        lonStaking.unstake();
+
+        vm.warp(block.timestamp + COOLDOWN_SECONDS + 1);
+
+        _redeemAndValidate(user, redeemAmount);
+    }
+
+    function _determineStakeRedeemAmount(uint80[2][16] memory stakeAndRedeemAmounts)
```

Comment:
Generating two different arrays, `stakeAmounts` and `redeemAomunts`, and making sure `stakeAmounts[i] > redeemAmounts[i]` take too much time so changed to just generating a pair of number and assigning bigger one to `stakeAmount`.